include/GafferImage/Constant.h

```diff
@@ -76,6 +76,7 @@ class GAFFERIMAGE_API Constant : public ImageNode
 
 		GafferImage::Format computeFormat( const Gaffer::Context *context, const ImagePlug *parent ) const override;
 		Imath::Box2i computeDataWindow( const Gaffer::Context *context, const ImagePlug *parent ) const override;
+		bool computeDeep( const Gaffer::Context *context, const ImagePlug *parent ) const override;
```

Comment:
The current convention is: it is not legitimate to pull sampleOffsets on a flat image.  In order to ensure that flat image processing nodes are not wasting time unnecessarily thinking about sample offsets, it always throws if you pull sample offsets on a flat image.  This does mean that nodes that do need to deal with sampleOffsets may need to do:
```

Reply:
> This seems like an acceptable convention to me.

Unless we have data to prove a performance benefit, I think there's a case for keeping things simple for ImagePlug clients, such that you can always call `sampleOffsetsPlug()->getValue()`. I'm working on the assumption that `FlatImageProcessor::hashSampleOffsets()` and `FlatImageProcessor::computeSampleOffsets()` could both be implemented in terms of `ImagePlug::flatTileSampleOffsets()`, with no need to pull on anything upstream. If that's the case, and if we disabled caching for that plug, I find it hard to imagine it being significantly slower than the special case code we're otherwise forcing onto all clients.

> Do I need to implement a more specific exceptions for calling sampleOffsets incorrectly? 

If we want to stick with the existing convention, then yes, I think this would help folks get things right. But since we'd probably throw the exception from `FlatImageProcessor::computeSampleOffsets()`, it doesn't seem like any more effort to just return `flatSampleOffsets()` instead.

> Also, the way that "NotImplementedException" is currently printed in currently quite confusing, because the type of exception isn't mentioned in the printout.

Indeed. It's getting lost when we wrap the original exception in ProcessException. I think I can sort that out - I'll take a look this morning.